### PR TITLE
Extends backend id to 32 bits

### PR DESCRIPTION
**Pull request by Weil0ng**
See commit msg.

This is not intended to be merged as is, the migrating/upgrading path is still missing. I would appreciate early feedbacks and will add migrating code later if needed.

This is sorta a stripped down version of PR #8041 to unblock us from running large-scale tests, we might as well bring back that original PR?

### CODE REVIEW

**Comment by Weil0ng** [comment]
I ran a large-scale test with this patch, seems to fix issue #16121

**Comment by aanm** [comment]
travis hit https://github.com/cilium/cilium/issues/16239

File: pkg/maps/lbmap/maglev.go [hunk 1]
```diff
@@ -201,7 +202,7 @@ func (k *MaglevInnerKey) String() string            { return fmt.Sprintf("%d", k
 // +k8s:deepcopy-gen=true
 // +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
 type MaglevInnerVal struct {
-	BackendIDs []uint16
+	BackendIDs []uint32
```
Comment (aanm) [inline pkg/maps/lbmap/maglev.go]: This should be `loadbalancer.BackendID`, no?

Comment (Weil0ng) [reply pkg/maps/lbmap/maglev.go]: It won't regenerate correct deepcopy method if it is `loadbalancer.BackendID` :)

**Comment by brb** [comment]
Marking it as draft due to missing upgrade/downgrade handling path.

@Weil0ng Are you planning to work on the upgrade/downgrade? Also, please refer the authors of the original PR in your commit msgs.

**Comment by Weil0ng** [comment]
> Marking it as draft due to missing upgrade/downgrade handling path.
> 
> @Weil0ng Are you planning to work on the upgrade/downgrade? Also, please refer the authors of the original PR in your commit msgs.

I'm thinking maybe we should just reuse your PR? any idea why it was reverted?

**Comment by Weil0ng** [comment]
> Marking it as draft due to missing upgrade/downgrade handling path.
> 
> @Weil0ng Are you planning to work on the upgrade/downgrade? Also, please refer the authors of the original PR in your commit msgs.

Also, just updated the commit msg with a reference to your original PR :) Should've add it in the first place, sry about that.

**Comment by brb** [comment]
> any idea why it was reverted?

It has been never merged. The reason for that is that the downgrade problem was not solved by the PR.

**Comment by Weil0ng** [comment]
> > any idea why it was reverted?
> 
> It has been never merged. The reason for that is that the downgrade problem was not solved by the PR:
> 
> > IMPORTANT However, the downgrade will interrupt the connections. Also, probably the manual cleanup of CT maps is required to avoid the wrong endpoint routing problem, as most likely different backend IDs will be allocated;-/ Alternatively, we could maintain both maps...

hmm, interesting, but is downgrade a requirement? I mean why would some one downgrade from 32 to 16 bits? to save memory?

**Comment by Weil0ng** [comment]
> Upon the downgrade we remove the u32 backend map and create the u16 map instead. What happens is that all backend ID allocations are lost => existing connections to services might be routed to wrong backends.

I see. In this case, it might be impossible to seamlessly downgrade, what if the original map (with u32) has more than 64k backends? then no matter how we migrate to u16, it is going to lose some backends?

**Comment by Weil0ng** [comment]
> I think we should introduce a hidden flag which would make both u16 and u32 to coexist and it would be used for the downgrade only.

So the downgrade (also upgrade?) process would be to switch on the hidden flag, wait until some condition is met (that u16 map has same number of entries as u32?), then downgrade/upgrade cilium version?

**Comment by brb** [comment]
My braindump:
- The new u32 will have a different name, e.g. `lb4_backends_v2`.
- The v1.11 datapath will refer to the new map.
- During the upgrade, the v1.11 cilium-agent will check whether the old map (`lb4_backends`) exists AND the new map is empty. If true, it will copy the entries from the old map to the new map.
- After the migration, if the proposed `--keep-legacy-service-backends` is set to false (default), then the old map is removed after we have reloaded the datapath for all endpoints (so that they all refer to the new map). If it is set to true, then we keep and maintain both maps. In this case, if backend ID > max(u16), then we just log a warning that downgrade might be disruptive.
- We document the flag in the upgrade guide.
- Once users are happy with the upgrade, they can set the flag to false, so that next time the map is removed.

**Comment by brb** [comment]
Another complication is that in the Maglev BPF maps we use u16 for backend ID.

**Comment by Weil0ng** [comment]
> Another complication is that in the Maglev BPF maps we use u16 for backend ID.

Is there a reason for that? or should we also change it to u32?

**Comment by brb** [comment]
> Is there a reason for that? or should we also change it to u32?

Yep, it should change to u32 too in a way that we don't make new connections to fail (existing connections should be fine, as backend ID is stored in CT).

**Comment by Weil0ng** [comment]
> My braindump:
> 
> * The new u32 will have a different name, e.g. `lb4_backends_v2`.
> * The v1.11 datapath will refer to the new map.
> * During the upgrade, the v1.11 cilium-agent will check whether the old map (`lb4_backends`) exists AND the new map is empty. If true, it will copy the entries from the old map to the new map.
> * After the migration, if the proposed `--keep-legacy-service-backends` is set to false (default), then the old map is removed after we have reloaded the datapath for all endpoints (so that they all refer to the new map). If it is set to true, then we keep and maintain both maps. In this case, if backend ID > max(u16), then we just log a warning that downgrade might be disruptive.
> * We document the flag in the upgrade guide.
> * Once users are happy with the upgrade, they can set the flag to false, so that next time the map is removed.

Coming back to this, doesn't cilium upgrade always reboot agent? If so, then agent will relist all services/backends upon rebooting and populate v2 map automatically, why do we need to explicitly copy from old map?

**Comment by varunmar** [comment]
What does the agent do on startup? When does it rewrite the eBPF programs on running pods?
If the existing agent rewrites the program before the maps are repopulated, we will need to keep the old maps around so there's no packet loss.

**Comment by brb** [comment]
> Coming back to this, doesn't cilium upgrade always reboot agent? If so, then agent will relist all services/backends upon rebooting and populate v2 map automatically, why do we need to explicitly copy from old map?

During the bootstrap cilium-agent gets all Service / Endpoint objects from k8s and only afterwards regenerates BPF programs of existing endpoints. My previous thinking for copying from v1 to v2 was to preserve the backend IDs (otherwise, we will route to wrong backends for existing connections). However, it should be enough to iterate over the v1 map and then restore the IDs. Just that we need to make sure that we don't leak IDs which can happen when a backend is removed before the upgrade while cilium-agent was down.

**Comment by Weil0ng** [comment]
> However, it should be enough to iterate over the v1 map and then restore the IDs. Just that we need to make sure that we don't leak IDs which can happen when a backend is removed before the upgrade while cilium-agent was down.

Still a bit confused. So v2 map has the up-to-date endpoint info from k8s (and it will be populated before BPF progs are regenerated), basically upon agent restart, we are good to go with v2 map, no? v1 map stores a snapshot before upgrade, so whatever happens (a backend removed/added) while cilium-agent is down should be captured when it syncs with apiserver and reflect in the v2 map? I don't follow why we could leak IDs in this way...

**Comment by brb** [comment]
>  So v2 map has the up-to-date endpoint info from k8s (and it will be populated before BPF progs are regenerated), basically upon agent restart, we are good to go with v2 map, no?

Please scratch the idea about restoring the IDs from v1, and then switching to v2. Unfortunately we still need to maintain both maps because of this:

>> * After the migration, if the proposed --keep-legacy-service-backends is set to false (default), then the old map is removed after we have reloaded the datapath for all endpoints (so that they all refer to the new map). If it is set to true, then we keep and maintain both maps. In this case, if backend ID > max(u16), then we just log a warning that downgrade might be disruptive.

**Comment by Weil0ng** [comment]
> * After the migration, if the proposed --keep-legacy-service-backends is set to false (default), then the old map is removed after we have reloaded the datapath for all endpoints (so that they all refer to the new map). If it is set to true, then we keep and maintain both maps. In this case, if backend ID > max(u16), then we just log a warning that downgrade might be disruptive.

Now I think about it, why do we need to maintain both, we will be using only one of them at a time? When you downgrade, cilium-agent would restart again and then repopulate u16 map with sync from k8s, no?

**Comment by brb** [comment]
> When you downgrade, cilium-agent would restart again and then repopulate u16 map with sync from k8s, no?

The backend IDs will be lost. We persist them only in the backends BPF map from which they are restored.

**Comment by Weil0ng** [comment]
> The backend IDs will be lost. We persist them only in the backends BPF map from which they are restored.

Yes. I think we should do this in two phases:
- in version X, we introduce v2 map silently and populate v2 map with whatever v1 map has, all operation still happen on v1 map, we just keep v2 map a duplicate of v1
- in version X+1, we switch all operation to occur with v2 map, and introduce `--keep-legacy-service-backends` flag as you suggested. If it is set to true, we keep populating v1 map when possible (for IDs > 64k, we issue warning), if it set to false, we drop v1 map completely.

So between current version and X, upgrade/downgrade will be seamless.
Between X and X+1, upgrade is seamless as v2 map already has all IDs in v1 map. Downgrade is seamless if `--keep-legacy-service-backends` was set to true and lossless if no backend ID is > 64k.

Does this make sense?

**Comment by Weil0ng** [comment]
Superseded by #17235. Closing.

**Pull request closed without merging**